.config/doom/config.org
#+title: My Custom Doom Emacs configuration
#+author: Harsh Vyapari (harshv5094)
#+date: <2024-10-25 Fri>
#+startup: showeverything

* Table Of Contents :toc:
- [[#basic-configuration][Basic Configuration]]
  - [[#clipboard][Clipboard]]
  - [[#doom-theme][Doom theme]]
  - [[#edit-specific-file][Edit Specific file]]
  - [[#font-family][Font Family]]
  - [[#global-auto-revert][Global Auto Revert]]
  - [[#lsp-delays][LSP Delays]]
  - [[#modeline][Modeline]]
  - [[#personal-information][Personal Information]]
  - [[#relative-line-number][Relative Line Number]]
  - [[#which-key][Which key]]
  - [[#window-splitting][Window splitting]]
- [[#markdown][Markdown]]
  - [[#markdown-headers][Markdown Headers]]
- [[#mouse-support][Mouse Support]]
- [[#org-mode-settings][Org Mode Settings]]
  - [[#org-agenda-default-file][Org agenda default file]]
  - [[#org-header-defaut-font-size][Org Header Defaut Font Size]]
  - [[#org-mode-directory-default-location][Org Mode Directory Default location]]
  - [[#org-journal-config][Org Journal Config]]

* Basic Configuration

** Clipboard
#+begin_src emacs-lisp :tangle yes
; NOTE: Setup clipboard
(setq select-enable-clipboard t)
#+end_src

** Doom theme
Setting default theme
#+begin_src emacs-lisp :tangle yes
; NOTE: Color Scheme
(setq doom-theme 'doom-gruvbox)
#+end_src

** Edit Specific file
Here we are adding keybinds to edit doom emacs config files
#+begin_src emacs-lisp :tangle yes
; NOTE: Custom doom emacs config keymaps
(map! :leader
      (:prefix ("=" . "open file")
       :desc "Edit agenda file"      "a" #'(lambda () (interactive) (find-file "~/org/agenda.org"))
       :desc "Edit doom config.org"  "c" #'(lambda () (interactive) (find-file "~/.config/doom/config.org"))
       :desc "Edit doom config.el"   "e" #'(lambda () (interactive) (find-file "~/.config/doom/config.el"))
       :desc "Edit doom init.el"     "i" #'(lambda () (interactive) (find-file "~/.config/doom/init.el"))
       :desc "Edit doom packages.el" "p" #'(lambda () (interactive) (find-file "~/.config/doom/packages.el"))))
#+end_src

** Font Family
Settings related to fonts within Doom Emacs:

+ ~doom-font~ - Standard monospace font that is used for most things in Emacs.
+ ~doom-variable-pitch-font~ Variable font which is useful in some Emacs plugin.
+ ~doom-big-font~ - Used in ~doom-big-font-mode~; useful for presentation.
+ ~font-lock-comment-face~ - for comments.
+ ~font-lock-keyword-face~ - for keywords with special significance like ~setq~ in elisp.

#+begin_src emacs-lisp :tangle yes
; NOTE: Setting Basic Font Family and Font Size
(setq doom-font (font-spec :family "JetBrainsMono Nerd Font" :size 15)
      doom-variable-pitch-font (font-spec :family "FiraCode Nerd Font Mono" :size 15)
      doom-big-font (font-spec :family "JetBrainsMono Nerd Font" :size 20))

; NOTE: Basic font style
(after! doom-themes
  (setq doom-themes-enable-bold t
        doom-themes-enable-italic t))

; NOTE: This are custom font faces for text
(custom-set-faces!
  '(font-lock-comment-face :slant italic))
  ;'(font-lock-keyword-face :slant italic))
#+end_src

** Global Auto Revert
A buffer can get out of sync with respect to its visited file on disk if that file is changed by another program. To keep it up to date, you can enable Auto Revert mode by typing M-x auto-revert-mode, or you can set it to be turned on globally with ‘global-auto-revert-mode’.  I have also turned on Global Auto Revert on non-file buffers, which is especially useful for ‘dired’ buffers.
#+begin_src emacs-lisp :tangle yes
; NOTE: global revert mode settings
(setq global-auto-revert-mode 1)
(setq global-auto-revert-non-file-buffers t)
#+end_src

** LSP Delays
#+begin_src emacs-lisp :tangle yes
; NOTE: add lsp delays
(after! flycheck (setq flycheck-idle-change-delay 0.1))
(after! lsp-mode
  (setq lsp-idle-delay 0.1)
  :custom
  (setq lsp-completion-enable-additional-text-edit t)
  (setq lsp-modeline-code-actions-enable t))
#+end_src

** Modeline
The modeline is the bottom status bar that appears in Emacs windows. For more information on what is available to configure in the Doom modeline, check out: https://github.com/seagle0128/doom-modeline
#+begin_src emacs-lisp :tangle yes
; NOTE: Customizing modeline design
(setq doom-modeline-height 30                    ;; Sets modeline height
      doom-modeline-bar-width 5                  ;; sets right bar width
      doom-modeline-buffer-file-name-style 'auto ;; auto setup doom modeline filename
      doom-modeline-persp-name t                 ;; adds perspective name to modeline
      doom-modeline-persp-icon t)                ;; adds folder icon next to persp name
#+end_src

** Personal Information
This contains various bits about myself that Emacs likes to use in various modules.
#+begin_src emacs-lisp :tangle yes
; NOTE: User information for Emacs to use
(setq user-full-name "Harsh Vyapari"
      user-mail-address "harshvy5094@proton.me")
#+end_src

** Relative Line Number
Enable relative line number
#+begin_src emacs-lisp :tangle yes
; NOTE: setting relative line number
(setq display-line-numbers-type 'relative)
#+end_src

** Which key
Reduce the popup delay of ~which-key~
#+begin_src emacs-lisp :tangle yes
; NOTE: Add which key delay for searching
(setq which-key-idle-delay 0.5)
#+end_src

** Window splitting
These changes make window splitting a bit nicer

First I want to focus to change to newly created windows.
#+begin_src emacs-lisp :tangle yes
; NOTE: Setting up split window rules for evil mode
(setq evil-vsplit-window-right t)
(setq evil-split-window-below t)
#+end_src

* Markdown

** Markdown Headers
Here we are giving markdown header some height.
#+begin_src emacs-lisp :tangle yes
; NOTE: Setting up custom markdown header
(custom-set-faces
 '(markdown-header-face ((t (:inherit font-lock-function-name-face :weight bold :family "FiraCode Nerd Font Mono"))))
 '(markdown-header-face-1 ((t (:inherit markdown-header-face :height 1.7))))
 '(markdown-header-face-2 ((t (:inherit markdown-header-face :height 1.6))))
 '(markdown-header-face-3 ((t (:inherit markdown-header-face :height 1.5))))
 '(markdown-header-face-4 ((t (:inherit markdown-header-face :height 1.4))))
 '(markdown-header-face-5 ((t (:inherit markdown-header-face :height 1.3))))
 '(markdown-header-face-6 ((t (:inherit markdown-header-face :height 1.2)))))
#+end_src

* Mouse Support
Adding the mouse support to the terminal version of the emacs
#+begin_src emacs-lisp :tangle yes
; NOTE: Enable mouse mode
(setq xterm-mouse-mode 1)
#+end_src

* Org Mode Settings

** Org agenda default file
Setting up org agenda default file name
#+begin_src emacs-lisp :tangle yes
; NOTE: Setting up org agenda file
(after! org
  (setq org-agenda-files '("~/org/agenda.org")))
#+end_src

** Org Header Defaut Font Size
Here is my custom function ~my-custom-header~ to change the font size of the org headings.

#+begin_src emacs-lisp :tangle yes
; NOTE Custom function to change header size
(custom-theme-set-faces! nil
'(org-level-8 :inherit outline-3 :height 1.0)
'(org-level-7 :inherit outline-3 :height 1.0)
'(org-level-6 :inherit outline-3 :height 1.1)
'(org-level-5 :inherit outline-3 :height 1.2)
'(org-level-4 :inherit outline-3 :height 1.3)
'(org-level-3 :inherit outline-3 :height 1.4)
'(org-level-2 :inherit outline-2 :height 1.5)
'(org-level-1 :inherit outline-1 :height 1.6)
'(org-document-title  :height 1.8 :bold t :underline nil))
#+end_src

** Org Mode Directory Default location
Setting different org directories default path default path
#+begin_src emacs-lisp :tangle yes
; NOTE Default Org Directory
(setq org-directory "~/org/")
; NOTE Default Note File
(setq org-default-notes-file (concat org-directory "/notes.org"))
#+end_src

** Org Journal Config
Setting Org Journal Default location
#+begin_src emacs-lisp :tangle yes
; NOTE Setting up org journal directory
(setq org-journal-dir "~/org/journal/"
      org-journal-date-prefix "* "
      org-journal-time-prefix "** "
      org-journal-date-format "%B %d, %Y (%A) "
      org-journal-file-format "%Y-%m-%d.org")
#+end_src


.config/doom/config.el
; NOTE: Setup clipboard
(setq select-enable-clipboard t)

; NOTE: Color Scheme
(setq doom-theme 'doom-gruvbox)

; NOTE: Custom doom emacs config keymaps
(map! :leader
      (:prefix ("=" . "open file")
       :desc "Edit agenda file"      "a" #'(lambda () (interactive) (find-file "~/org/agenda.org"))
       :desc "Edit doom config.org"  "c" #'(lambda () (interactive) (find-file "~/.config/doom/config.org"))
       :desc "Edit doom config.el"   "e" #'(lambda () (interactive) (find-file "~/.config/doom/config.el"))
       :desc "Edit doom init.el"     "i" #'(lambda () (interactive) (find-file "~/.config/doom/init.el"))
       :desc "Edit doom packages.el" "p" #'(lambda () (interactive) (find-file "~/.config/doom/packages.el"))))

; NOTE: Setting Basic Font Family and Font Size
(setq doom-font (font-spec :family "JetBrainsMono Nerd Font" :size 15)
      doom-variable-pitch-font (font-spec :family "FiraCode Nerd Font Mono" :size 15)
      doom-big-font (font-spec :family "JetBrainsMono Nerd Font" :size 20))

; NOTE: Basic font style
(after! doom-themes
  (setq doom-themes-enable-bold t
        doom-themes-enable-italic t))

; NOTE: This are custom font faces for text
(custom-set-faces!
  '(font-lock-comment-face :slant italic))
  ;'(font-lock-keyword-face :slant italic))

; NOTE: global revert mode settings
(setq global-auto-revert-mode 1)
(setq global-auto-revert-non-file-buffers t)

; NOTE: add lsp delays
(after! flycheck (setq flycheck-idle-change-delay 0.1))
(after! lsp-mode
  (setq lsp-idle-delay 0.1)
  :custom
  (setq lsp-completion-enable-additional-text-edit t)
  (setq lsp-modeline-code-actions-enable t))

; NOTE: Customizing modeline design
(setq doom-modeline-height 30                    ;; Sets modeline height
      doom-modeline-bar-width 5                  ;; sets right bar width
      doom-modeline-buffer-file-name-style 'auto ;; auto setup doom modeline filename
      doom-modeline-persp-name t                 ;; adds perspective name to modeline
      doom-modeline-persp-icon t)                ;; adds folder icon next to persp name

; NOTE: User information for Emacs to use
(setq user-full-name "Harsh Vyapari"
      user-mail-address "harshvy5094@proton.me")

; NOTE: setting relative line number
(setq display-line-numbers-type 'relative)

; NOTE: Add which key delay for searching
(setq which-key-idle-delay 0.5)

; NOTE: Setting up split window rules for evil mode
(setq evil-vsplit-window-right t)
(setq evil-split-window-below t)

; NOTE: Setting up custom markdown header
(custom-set-faces
 '(markdown-header-face ((t (:inherit font-lock-function-name-face :weight bold :family "FiraCode Nerd Font Mono"))))
 '(markdown-header-face-1 ((t (:inherit markdown-header-face :height 1.7))))
 '(markdown-header-face-2 ((t (:inherit markdown-header-face :height 1.6))))
 '(markdown-header-face-3 ((t (:inherit markdown-header-face :height 1.5))))
 '(markdown-header-face-4 ((t (:inherit markdown-header-face :height 1.4))))
 '(markdown-header-face-5 ((t (:inherit markdown-header-face :height 1.3))))
 '(markdown-header-face-6 ((t (:inherit markdown-header-face :height 1.2)))))

; NOTE: Enable mouse mode
(setq xterm-mouse-mode 1)

; NOTE: Setting up org agenda file
(after! org
  (setq org-agenda-files '("~/org/agenda.org")))

; NOTE Custom function to change header size
(custom-theme-set-faces! nil
'(org-level-8 :inherit outline-3 :height 1.0)
'(org-level-7 :inherit outline-3 :height 1.0)
'(org-level-6 :inherit outline-3 :height 1.1)
'(org-level-5 :inherit outline-3 :height 1.2)
'(org-level-4 :inherit outline-3 :height 1.3)
'(org-level-3 :inherit outline-3 :height 1.4)
'(org-level-2 :inherit outline-2 :height 1.5)
'(org-level-1 :inherit outline-1 :height 1.6)
'(org-document-title  :height 1.8 :bold t :underline nil))

; NOTE Default Org Directory
(setq org-directory "~/org/")
; NOTE Default Note File
(setq org-default-notes-file (concat org-directory "/notes.org"))

; NOTE Setting up org journal directory
(setq org-journal-dir "~/org/journal/"
      org-journal-date-prefix "* "
      org-journal-time-prefix "** "
      org-journal-date-format "%B %d, %Y (%A) "
      org-journal-file-format "%Y-%m-%d.org")
